src/runtime-rs/crates/resource/src/cpu_mem/mem.rs

```diff
@@ -0,0 +1,171 @@
+            container_mem_resources: Arc::new(RwLock::new(HashMap::new())),
+            use_guest_swap: hypervisor_config.memory_info.enable_guest_swap,
+            orig_toml_default_mem: init_size_manager.get_orig_toml_default_mem(),
+        })
+    }
+
+    pub(crate) async fn update_mem_resources(
+        &self,
+        cid: &str,
+        linux_resources: Option<&LinuxResources>,
+        op: ResourceUpdateOp,
+        hypervisor: &dyn Hypervisor,
+    ) -> Result<()> {
+        self.update_container_mem_resources(cid, linux_resources, op)
+            .await
+            .context("update container memory resources")?;
+        // the unit here is byte
+        let (mem_sb_byte, need_pod_swap, swap_sb_byte) = self
+            .total_mems(self.use_guest_swap)
+            .await
+            .context("failed to calculate total memory requirement for containers")?;
+        let mut mem_sb_mb = (mem_sb_byte >> MIB_TO_BYTES_SHIFT) as u32;
+        mem_sb_mb += self.orig_toml_default_mem;
+        if need_pod_swap {
+            let swap_sb_mb = (swap_sb_byte >> MIB_TO_BYTES_SHIFT) as u32;
+            mem_sb_mb += swap_sb_mb;
+        }
+        info!(sl!(), "calculate mem_sb_mb {}", mem_sb_mb);
+
+        let curr_mem = self
+            .do_update_mem_resource(mem_sb_mb, swap_sb_byte, hypervisor)
+            .await
+            .context("failed to update_mem_resource")?;
+
+        self.update_current_mem(curr_mem).await;
+        Ok(())
+    }
+
+    async fn update_current_mem(&self, new_mem: u32) {
+        let mut current_mem = self.current_mem.write().await;
+        *current_mem = new_mem;
+    }
+
+    async fn get_current_mem(&self) -> u32 {
+        let current_mem = self.current_mem.read().await;
+        *current_mem
+    }
+
+    async fn total_mems(&self, use_guest_swap: bool) -> Result<(u64, bool, i64)> {
+        // sb stands for sandbox
+        let mut mem_sandbox = 0;
+        let mut need_pod_swap = false;
+        let mut swap_sandbox = 0;
+
+        let resources = self.container_mem_resources.read().await;
+
+        for (_, r) in resources.iter() {
+            for l in &r.hugepage_limits {
+                mem_sandbox += l.limit;
+            }
+
+            if let Some(memory) = &r.memory {
+                let current_limit = memory.limit.map_or(0, |limit| {
+                    mem_sandbox += limit as u64;
+                    info!(sl!(), "memory sb: {}, memory limit: {}", mem_sandbox, limit);
+                    limit
+                });
+
+                if let Some(swappiness) = memory.swappiness {
+                    if swappiness > 0 && use_guest_swap {
+                        if let Some(swap) = memory.swap {
+                            if swap > current_limit {
+                                swap_sandbox = swap.saturating_sub(current_limit);
+                            }
+                        } else if current_limit == 0 {
+                            need_pod_swap = true;
```

Comment:
1. Why `need_pod_swap = true;` when the `current_limit == 0`, suggest adding some comments.
2. What about multi-container but only one container `current_limit == 0`.

Reply:
1. if `current_limit` is 0, the container will have access to the entire memory available on the host system so that we will enable swap in the guest.
2. `need_pod_swap` is a global variable for this function, if only one `current_limit == 0`, we still need to enable swap

Reply:
Seems `-1` means `unlimited`, not `0`
ref: https://github.com/opencontainers/runtime-spec/blob/main/config-linux.md#memory